handler/src/main/java/io/netty/handler/ssl/KeyManagerConfiguration.java

```diff
@@ -0,0 +1,172 @@
+
+import java.io.File;
+import java.io.InputStream;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Collections;
+
+/**
+ * Configures the KeyManager for an SslContextBuilder.
+ *
+ * If the instance is not for configuring a server, key and keyCertChain may be left unspecified,
+ * which disables mutual authentication.
+ *
+ * If the key is not password protected, the password may be left unspecified.
+ */
+public final class KeyManagerConfiguration {
+
+    private PrivateKey key;
+    private X509Certificate[] keyCertChain;
+    private String keyPassword;
+
+    /**
+     * Sets the private key for this host.
+     *
+     * @param key a PKCS#8 private key
+     * @param keyPassword the password of the key, or null if it's not password protected
+     */
+    public KeyManagerConfiguration key(PrivateKey key, String keyPassword) {
+        this.key = key;
+        this.keyPassword = keyPassword;
+        return this;
+    }
+
+    /**
+     * Sets the private key for this host.
+     *
+     * @param key a PKCS#8 private key file in PEM format
+     * @param keyPassword the password of the key, or null if it's not password protected
+     */
+    public KeyManagerConfiguration key(File keyFile, String keyPassword) {
+        try {
+            return key(SslContext.toPrivateKey(keyFile, keyPassword), keyPassword);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("File does not contain valid private key: " + keyFile, e);
+        }
+    }
+
+    /**
+     * Sets the private key for this host.
+     *
+     * @param key a PKCS#8 private key input stream in PEM format
+     * @param keyPassword the password of the key, or null if it's not password protected
+     */
+    public KeyManagerConfiguration key(InputStream keyInputStream, String keyPassword) {
+        try {
+            return key(SslContext.toPrivateKey(keyInputStream, keyPassword), keyPassword);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Input stream does not contain valid private key.", e);
+        }
+    }
+
+    /**
+     * Retrieves the private key for this host.
+     *
+     * @note Should not be invoked before calling {@code validate}
+     */
+    PrivateKey getKey() {
+        return key;
+    }
+
+    /**
+     * Retrieves the password for the private key for this host.
+     *
+     * @note Should not be invoked before calling {@code validate}
+     */
+    String getPassword() {
+        return keyPassword;
+    }
+
+    /**
+     * Sets the X.509 certificate chain.
+     *
+     * @param keyCertChain an X.509 certificate chain
+     */
+    public KeyManagerConfiguration keyCertChain(X509Certificate... keyCertChain) {
+        if (keyCertChain == null || keyCertChain.length == 0) {
+            this.keyCertChain = null;
+        } else {
+            this.keyCertChain = keyCertChain.clone();
```

Comment:
we don't need to clone if we have already created a new intermediate array from one of the other `keyCertChain` methods. Consider adding an internal method which avoids the clone if called internally.